File: app/src/main/java/sgtmelon/scriptum/app/model/item/RollItem.java
package sgtmelon.scriptum.app.model.item;

import org.json.JSONException;
import org.json.JSONObject;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.ForeignKey;
import androidx.room.Index;
import androidx.room.PrimaryKey;
import androidx.room.TypeConverters;
import sgtmelon.scriptum.app.model.NoteRepo;
import sgtmelon.scriptum.office.annot.DbAnn;
import sgtmelon.scriptum.office.conv.BoolConv;

import static androidx.room.ForeignKey.CASCADE;

/**
 * Элемент списка заметок {@link NoteRepo}
 */
@Entity(tableName = DbAnn.Roll.TABLE,
        foreignKeys = @ForeignKey(entity = NoteItem.class,
                parentColumns = DbAnn.Note.ID,
                childColumns = DbAnn.Roll.NOTE_ID,
                onUpdate = CASCADE,
                onDelete = CASCADE),
        indices = {@Index(DbAnn.Roll.NOTE_ID)})
@TypeConverters({BoolConv.class})
public final class RollItem {

    @ColumnInfo(name = DbAnn.Roll.ID)
    @PrimaryKey(autoGenerate = true)
    @Nullable
    private Long id;

    @ColumnInfo(name = DbAnn.Roll.NOTE_ID) private long noteId;
    @ColumnInfo(name = DbAnn.Roll.POSITION) private int position;
    @ColumnInfo(name = DbAnn.Roll.CHECK) private boolean check = false;
    @ColumnInfo(name = DbAnn.Roll.TEXT) private String text;

    public RollItem() {

    }

    public RollItem(String data) {
        try {
            final JSONObject jsonObject = new JSONObject(data);

            id = jsonObject.getLong(DbAnn.Roll.ID);
            noteId = jsonObject.getLong(DbAnn.Roll.NOTE_ID);
            position = jsonObject.getInt(DbAnn.Roll.POSITION);
            check = jsonObject.getBoolean(DbAnn.Roll.CHECK);
            text = jsonObject.getString(DbAnn.Roll.TEXT);
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    @Nullable
    public Long getId() {
        return id;
    }

    public void setId(@NonNull Long id) {
        this.id = id;
    }

    public long getNoteId() {
        return noteId;
    }

    public void setNoteId(long noteId) {
        this.noteId = noteId;
    }

    public int getPosition() {
        return position;
    }

    public void setPosition(int position) {
        this.position = position;
    }

    public boolean isCheck() {
        return check;
    }

    public void setCheck(boolean check) {
        this.check = check;
    }

    @NonNull
    public String getText() {
        return text;
    }

    public void setText(@NonNull String text) {
        this.text = text;
    }

    @NonNull
    @Override
    public String toString() {
        final JSONObject jsonObject = new JSONObject();

        try {
            jsonObject.put(DbAnn.Roll.ID, id);
            jsonObject.put(DbAnn.Roll.NOTE_ID, noteId);
            jsonObject.put(DbAnn.Roll.POSITION, position);
            jsonObject.put(DbAnn.Roll.CHECK, check);
            jsonObject.put(DbAnn.Roll.TEXT, text);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        return jsonObject.toString();
    }

}

File: app/src/main/java/sgtmelon/scriptum/model/item/RollItem.kt
package sgtmelon.scriptum.model.item

import androidx.room.ColumnInfo
import androidx.room.TypeConverters
import org.json.JSONObject
import sgtmelon.scriptum.adapter.RollAdapter
import sgtmelon.scriptum.model.data.DbData.Roll
import sgtmelon.scriptum.model.item.RollItem.Companion.get
import sgtmelon.scriptum.room.converter.BoolConverter

/**
 * Model for store short information about roll, use in [RollAdapter]
 */
@TypeConverters(BoolConverter::class)
data class RollItem(
        @ColumnInfo(name = Roll.ID) var id: Long? = Roll.Default.ID,
        @ColumnInfo(name = Roll.POSITION) var position: Int,
        @ColumnInfo(name = Roll.CHECK) var isCheck: Boolean = Roll.Default.CHECK,
        @ColumnInfo(name = Roll.TEXT) var text: String
) {
    /**
     * Replace [id] null value to -1 for [get] function
     */
    fun toJson() = JSONObject().apply {
        put(Roll.ID, if (id != null) id else -1L)
        put(Roll.POSITION, position)
        put(Roll.CHECK, isCheck)
        put(Roll.TEXT, text)
    }.toString()

    companion object {
        operator fun get(data: String): RollItem? = try {
            JSONObject(data).let {
                val id = it.getLong(Roll.ID)
                return@let RollItem(
                        if (id != -1L) id else null,
                        it.getInt(Roll.POSITION),
                        it.getBoolean(Roll.CHECK),
                        it.getString(Roll.TEXT)
                )
            }
        } catch (e: Throwable) {
            null
        }
    }

}